airflow/contrib/hooks/ssh_hook.py

```diff
@@ -14,125 +14,141 @@
-            connection_cmd += ["-o", "UserKnownHostsFile=/dev/null",
-                               "-o", "StrictHostKeyChecking=no"]
-
-        if self.key_file:
-            connection_cmd += ["-i", self.key_file]
-
-        if self.tty:
-            connection_cmd += ["-t"]
-
-        connection_cmd += cmd
-        logging.debug("SSH cmd: {} ".format(connection_cmd))
-
-        return connection_cmd
-
-    def Popen(self, cmd, **kwargs):
-        """
-        Remote Popen
-
-        :param cmd: command to remotely execute
-        :param kwargs: extra arguments to Popen (see subprocess.Popen)
-        :return: handle to subprocess
-        """
-        prefixed_cmd = self._prepare_command(cmd)
-        return subprocess.Popen(prefixed_cmd, **kwargs)
-
-    def check_output(self, cmd):
-        """
-        Executes a remote command and returns the stdout a remote process.
-        Simplified version of Popen when you only want the output as a string and detect any errors.
-
-        :param cmd: command to remotely execute
-        :return: stdout
-        """
-        p = self.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-        output, stderr = p.communicate()
-
-        if p.returncode != 0:
-            # I like this better: RemoteCalledProcessError(p.returncode, cmd, self.host, output=output)
-            raise AirflowException("Cannot execute {} on {}. Error code is: {}. Output: {}, Stderr: {}".format(
-                                   cmd, self.conn.host, p.returncode, output, stderr))
-
-        return output
+    def get_conn(self):
+        if not self.client:
+            logging.debug('creating ssh client for conn_id: {0}'.format(self.ssh_conn_id))
+            if self.ssh_conn_id is not None:
+                conn = self.get_connection(self.ssh_conn_id)
+                if self.username is None:
+                    self.username = conn.login
+                if self.password is None:
+                    self.password = conn.password
+                if self.remote_host is None:
+                    self.remote_host = conn.host
+                if conn.extra is not None:
+                    extra_options = conn.extra_dejson
+                    self.key_file = extra_options.get("key_file")
+
+                    if "timeout" in extra_options:
+                        self.timeout = int(extra_options["timeout"], 10)
+
+                    if "compress" in extra_options \
+                            and extra_options["compress"].lower() == 'false':
+                        self.compress = False
+                    if "no_host_key_check" in extra_options \
+                            and extra_options["no_host_key_check"].lower() == 'false':
+                        self.compress = False
+
+            if not self.remote_host:
+                raise AirflowException("Missing required param: remote_host")
+
+            # Auto detecting some values from system
+            if not self.username:
```

Comment:
so I added this based on the thought that this is actually the last try to get username after trying to get it from Connection and params, and will fail if its not intended to, but gives a little convenience so that username doesn't has to be provided in cases where the airflow process is running as intended user.